File: src/main/java/ac/dnd/server/admission/infrastructure/persistence/crypto/HmacBlindIndexCreator.java
package ac.dnd.server.admission.infrastructure.persistence.crypto;

import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

import ac.dnd.server.common.propeties.EncryptionProperties;

public class HmacBlindIndexCreator {

	private static final String HMAC_ALGORITHM = "HmacSHA256";
	private final EncryptionProperties properties;

	public HmacBlindIndexCreator(
		final EncryptionProperties properties
	) {
		this.properties = properties;
	}

	public String create(
		final String plainText
	) {
		try {
			Mac mac = Mac.getInstance(HMAC_ALGORITHM);
			SecretKeySpec secretKeySpec = new SecretKeySpec(
				properties.getHmac().getKey()
					.getBytes(StandardCharsets.UTF_8),
				HMAC_ALGORITHM
			);

			mac.init(secretKeySpec);
			byte[] hmacBytes = mac.doFinal(plainText.getBytes(StandardCharsets.UTF_8));
			return Base64.getEncoder().encodeToString(hmacBytes);

		} catch (NoSuchAlgorithmException | InvalidKeyException e) {
			throw new RuntimeException(
				"Failed to create blind index",
				e
			);
		}
	}
}

File: src/main/kotlin/ac/dnd/server/admission/infrastructure/persistence/crypto/HmacBlindIndexCreator.kt
package ac.dnd.server.admission.infrastructure.persistence.crypto

import ac.dnd.server.shared.config.properties.EncryptionProperties
import java.nio.charset.StandardCharsets
import java.security.InvalidKeyException
import java.security.NoSuchAlgorithmException
import java.util.*
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

class HmacBlindIndexCreator(
    private val properties: EncryptionProperties
) {
    companion object {
        private const val HMAC_ALGORITHM = "HmacSHA256"
    }

    private val secretKeySpec: SecretKeySpec = SecretKeySpec(
        properties.hmac.key.toByteArray(StandardCharsets.UTF_8),
        HMAC_ALGORITHM
    )

    fun create(plainText: String): String {
        return try {
            val mac = Mac.getInstance(HMAC_ALGORITHM)
            mac.init(secretKeySpec)
            val hmacBytes = mac.doFinal(plainText.toByteArray(StandardCharsets.UTF_8))
            Base64.getEncoder().encodeToString(hmacBytes)
        } catch (e: NoSuchAlgorithmException) {
            throw RuntimeException("Failed to create blind index", e)
        } catch (e: InvalidKeyException) {
            throw RuntimeException("Failed to create blind index", e)
        }
    }
}
